### ISSUE TITLE
Reproducible Builds

### ISSUE THREAD

**Issue by IzzySoft** [sidhant947/Wallet#1]
Eh, had the wrong text in here, too many open tabs :see_no_evil: But while I'm here, and the issue was already created, let me transfer this over here:

Strange, today's release was no longer reproducible:

```diff
  -rw-r--r--  0.0 unx   743616 b-   307390 defN 1981-01-01 01:01:02 1c06cd94 classes.dex
- -rw-r--r--  0.0 unx  5309344 b-  2119722 defN 1981-01-01 01:01:02 90369b6e lib/arm64-v8a/libapp.so
+ -rw-r--r--  0.0 unx  5309344 b-  2119724 defN 1981-01-01 01:01:02 7bbc2377 lib/arm64-v8a/libapp.so
  -rw-r--r--  0.0 unx 10714752 b-  4991251 defN 1981-01-01 01:01:02 4a340eb2 lib/arm64-v8a/libflutter.so
```

Almost looks as if you had compiled that on Fedora: the **uncompressed** `libapp.so` are identical (4th column: uncompressed size), but they are compressed differently (6th column: compressed size), resulting in a different hash (10th column). Same compression level, but different results – as if at your end it was not `zlib`, but `zlib-ng` (known issue). Could that be? Added to that, you forgot to [increase `versionCode`](https://github.com/sidhant947/Wallet/blob/main/pubspec.yaml#L18) (the "+1" at the end of that line should meanwhile be a `3`, as there were 2 previous releases already), so your reproducible version here was overwritten by the non-RB one:

    Binary files /web/ftp/repo/fdroid/repo/com.sidhant.wallet_2001.apk and /web/ftp/repo/fdroid/repo/com.sidhant.wallet_1.0.2.apk differ

That also means that whoever had installed v1.0.0 or v1.0.1 will never receive this update (Android checks `versionCode` and only signals "new version" if it was increased).

So 3 things for you, @sidhant947:

* make sure to always increase `versionCode` when you make a new release (adjusting `versionName` does not suffice)
* we need to find out what causes this compression mismatch. What did you change on your system between the previous build and this one?
* finally, that all fixed, we'll need a new release to clean up this mess :see_no_evil:

**Comment by sidhant947** [comment]
Sorry About VersionCode Mistake. I was busy in Google Fraud Play Store to get beta Testers , it skipped my mind to update the later part. I promise it won't happen again. Let me know if code is lacking somewhere else too , I'll try my best to fix it.

No I havn't changed my system , Build are from same dir as previous ones.

`/home/sidhant/Downloads/Projects/wallet/wallet/<Project code>`

**Comment by IzzySoft** [comment]
Mistakes happen (that were those things causing "ah, deja vu" effects when doing them again, right? :speak_no_evil: :dash:) And yeah, nowadays one better leaves PlayStore out.

If you didn't change anything at your end I wonder what caused that compression fluke. I can of course try building again and see if the results are different (sometimes builds are flaky) – but I doubt it (update: my doubt was confirmed). [That `zlib-ng` issue](https://lists.reproducible-builds.org/pipermail/rb-general/2024-September/003547.html) to my knowledge so far only effects Fedora, but Arch ships it as well. You sure you didn't accidentally switch to that? What JDK are you using? The work-around on Fedora is to stick with OpenJDK (which bundles its own `zlib`, not ng).

While here, you still didn't add the flutter version to your `pubspec.yaml`:

```yaml
environment:
  sdk: ^3.5.4
  flutter: ^3.24.5
```

Can you do that (and keep it aligned with what you use)? That helps keeping RB recipes aligned as the Flutter version then could be auto-adjusted in the recipe.

**Comment by sidhant947** [comment]
JDK is 22.0.2 , I have not changed directory
![Screenshot from 2024-11-23 08-20-46](https://github.com/user-attachments/assets/11578508-c92e-41ad-b8e5-c6fc05326071)
![Screenshot from 2024-11-23 08-23-15](https://github.com/user-attachments/assets/01b77188-6a75-4c2f-b823-a3e853257a8b)

**Comment by IzzySoft** [comment]
> JDK is 22.0.2

Oof, that's not a stable release (11, 17 and 21 are the current stable ones). Have the previous releases also been built with 22? Could you switch to 21 (until the next stable becomes available at least)?

**Comment by sidhant947** [comment]
Downgraded , Will Stay on stable build until next stable release
![Screenshot from 2024-11-23 22-14-04](https://github.com/user-attachments/assets/3536946b-6404-4d08-8f2a-e00897b81d27)

Previous releases were build on java 17 lts i guess , build.gradle have this 
`  compileOptions {
        sourceCompatibility = JavaVersion.VERSION_17
        targetCompatibility = JavaVersion.VERSION_17
    }

    kotlinOptions {
        jvmTarget = JavaVersion.VERSION_17
    }`

**Comment by IzzySoft** [comment]
> Downgraded , Will Stay on stable build until next stable release

Thanks! Any ETA for the next release? Or do you want to build an APK from the currently latest commit for me to test?

> Previous releases were build on java 17

At least here they were. And AFAIK `sourceCompatibility` just specifies the minimum JDK required (I have several apps in my builder having that set to `8` but compiling fine with 11 or 17). I'm also not sure what effect `jvmTarget` has, had no time yet to dig into that. But there we go:

* [`jvmTarget`](https://kotlinlang.org/docs/gradle-compiler-options.html#attributes-specific-to-jvm): "Target version of the generated JVM bytecode". They don't explain, but if I understand correctly it means "generate byte code compatible to that Java version" – which implies you must have **at least** that version available, as a lower one would not know how to do that
* `sourceCompatibility`: code needs at least this Java version to compile
* `targetCompatibility`: resulting app needs at least that Java version (cannot run with a lower JRE)

So all 3 just define a minimum, not an absolute version.

**Comment by IzzySoft** [comment]
OK, I found the culprit using https://try.diffoscope.org/

![image](https://github.com/user-attachments/assets/0d37c260-0045-4eb3-b821-7c3cfd5cd318)

You now use a buildID somewhere it seems. But I'm not familiar enough with Flutter to tell where you might have done that, and how you could fix that. As you obviously didn't have that with v1.0.1, it must be something you've changed recently. Searching the repo, I did not get a single hit on ["buildid"](https://api.flutter.dev/flutter/dart-developer/NativeRuntime/buildId.html) – but maybe it rings a bell with you?

BuildIds are often a cause for RBs to fail, unless they are explicitly made deterministic, see e.g. [here in our wiki](https://gitlab.com/IzzyOnDroid/repo/-/wikis/Reproducible-Builds#no-funny-build-time-generated-ids).

**Comment by sidhant947** [comment]
It may be associated with versionCode mistake i did in previous build , Released new Version , Try if it solves error now

**Comment by IzzySoft** [comment]
No, those two are unreleated. And no, V1.0.3 is not RB, still the same issue with the buildid:

![image](https://github.com/user-attachments/assets/501445b6-6b12-412c-a2fc-8194c7a0a948)

If you don't remember having added that, maybe some new dependency you've added did that? I see e.g. `syncfusion_flutter_barcodes` and `local_auth` haven't been there with v1.0.1. But I'm no flutter dev, I cannot much help finding out. But I found a [related issue](https://github.com/dart-lang/sdk/issues/52506) – which raises the question of the NDK location. According to the settings in your `android/app/build.gradle` you're using `25.1.8937393`, so on my end the NDK is in `/opt/sdk/ndk/25.1.8937393`. Which one are you using de-facto?

**Comment by sidhant947** [comment]
i don't think it is due to syncfusion / local_auth as issue was in 1.0.2 also , Look like a version code issue as Ndk version havn't changed in previous builds , It's Issue in due to version: 1.0.2+1 this line i think as it wasn't updated in V1.0.1 also , so that seem to be issue as Not proper Build ID management from start as it should have been version: 1.0.1+2 in V1.0.1 as it did not gave any error in that , otherwise i havn't changed anything other than BioMetric Auth in app since V1.0..1 , should i try using Diffrent config like v1.0.3+3 maybe that can solve the issue

**Comment by IzzySoft** [comment]
Again, @sidhant947: build-ids have nothing to do with `versionCode` (or it would reproduce at my end as the same is used here: it's a fixed numbers and thus deterministic). BuildIDs are generated either explicitly, or (in case of Flutter) implicitly depending on build and NDK paths (see the linked issue – that's why I asked for which NDK version you have installed and in which path, maybe something there had changed when you upgraded from JDK-17 to JDK-22). But I'm not familiar enough with Flutter to debug that.

**Comment by sidhant947** [comment]
I'm on NDK LTS version "27.2.12479018" , I'm also not familiar with debugging Flutter that much , I have never encountered this error of Build ID.

**Comment by IzzySoft** [comment]
OK, specifying that NDK version directly changed something, though I'm not sure what it is now:
[diff.zip](https://github.com/user-attachments/files/17895983/diff.zip)

This is what diffoscope sees as differences in `libapp.so` for V1.0.3 now:

![image](https://github.com/user-attachments/assets/21fe40dd-aaf6-4625-9339-a978bf06a3c2)

But I've no clue what to make of it, or how to go from here (maybe someone more knowable can give us a clue). APK was built with Flutter 3.24.5, JDK-21 and NDK 27.2.12479018 (`/opt/sdk/ndk/27.2.12479018`).

**Comment by IzzySoft** [comment]
@sidhant947 any news on this? Pops up at each daily run here. It's been over a week now, usually I keep such "fails" back for max 1..2 days. If I cannot fix the build on our end, I'll have to show it as "failed" with the next run in about 8h from now – and hope we can get it tackled for the next release then.

I don't know what (apart from the build location) is used to generate the flutter build-id; but maybe you can provide the output of `flutter doctor -v` so we can compare? Not sure how much that helps, but might be worth a try.

**Comment by sidhant947** [comment]
Flutter (Channel stable, 3.24.5, on Manjaro Linux 6.11.10-2-MANJARO, locale en_US.UTF-8)
    • Flutter version 3.24.5 on channel stable at /home/sidhant/Softwares/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision dec2ee5c1f (3 weeks ago), 2024-11-13 11:13:06 -0800
    • Engine revision a18df97ca5
    • Dart version 3.5.4
    • DevTools version 2.37.3

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /home/sidhant/Android/Sdk
    • Platform android-34, build-tools 34.0.0
    • Java binary at: /home/sidhant/Softwares/android-studio/jbr/bin/java
    • Java version OpenJDK Runtime Environment (build 21.0.3+-12282718-b509.11)
    • All Android licenses accepted.

[✗] Chrome - develop for the web (Cannot find Chrome executable at google-chrome)
    ! Cannot find Chrome. Try setting CHROME_EXECUTABLE to a Chrome executable.

[✓] Linux toolchain - develop for Linux desktop
    • clang version 18.1.8
    • cmake version 3.31.1
    • ninja version 1.12.1
    • pkg-config version 2.1.1

[✓] Android Studio (version 2024.2)
    • Android Studio at /home/sidhant/Softwares/android-studio
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 21.0.3+-12282718-b509.11)

[✓] Connected device (1 available)
    • Linux (desktop) • linux • linux-x64 • Manjaro Linux 6.11.10-2-MANJARO

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.

**Comment by sidhant947** [comment]
I reached out for this into the flutter community, surprisingly a 6year experience developer just give up his hands and not getting response from reddit also 🥲 i thought of it as minor bug but nobody is giving any clue . Maybe i should change and start a new build in a new tree , lost all hope for flutter, not getting any support from their devs community also

**Comment by IzzySoft** [comment]
One failing RB isn't the end of the world, @sidhant947 – we could wait and see if we have more luck with the next release, taking care for building from a clean tree and all. As pointed out, those `build-id`s are calculated at build time, and take things like build path as well as flutter versions (and some more which escapes me) into account, so we need to match them all. Rust has a similar issue. The only ones where I haven't encountered such yet were pure Java or pure Kotlin apps (with no "native libs" – with the latter, there's sometimes a similar issue concerning the versions of their build tools).

And don't you think switching to Expo: that's often especially hard to get RB…

> i thought of it as minor bug

Well, I've linked you a [related issue](https://github.com/dart-lang/sdk/issues/52506), that sums up parts of what goes into that build-id. Which is what we partially walked by. When setting up flutter apps for RB, we always use the same build path (no use of even trying without). I then tried to match the NDK path, too.

The only thing left then could be a not-pinned version of a library/flutter-plugin. As far as I can tell, all the plugins you've listed in `pubspec.yaml` are locked to their versions (but it might also be "recursion hell", i.e. one of those libraries requiring another one, idk).

Wanted to give your new v1.0.4 a try, but that will not only fail RB but also its update in the repo later:

    Error: APK appid mismatch: expected com.sidhant.wallet, got com.sidhant.chill_guy

A different appid means a different app, so this no longer matches **anything** here now. I had to disable updates in the repo as well. There's no way to update from 1.0.3 to 1.0.4 with the apps having different IDs. I wonder how this could happen, as `com.sidhant.chill_guy` cannot be found in this repo – and even the `build.gradle` at the tagged release uses `com.sidhant.wallet`?

    package: name='com.sidhant.chill_guy' versionCode='2001' versionName='0.1.0'

Looks like you've attached the APKs from [this repo](https://github.com/sidhant947/chill_guy)?

**Comment by sidhant947** [comment]
look like  i'm embarassing myself with every new day in front of you

**Comment by IzzySoft** [comment]
Eh, :hankey: happens all the time – we are humans, and to err is human. Only to screw up mightily it needs a computer… wait, **that's** the problem, we're using computers! :see_no_evil:

Take it positively: **this** issue was easy to solve, wasn't it? :stuck_out_tongue_winking_eye:

**Comment by sidhant947** [comment]
Well someone told me to  completely analyze the v1.0.1 which did not had any error to v1.0.2 , i saw this 

 dependenciesInfo {
        // Disables dependency metadata when building APKs.
        includeInApk = false
        // Disables dependency metadata when building Android App Bundles.
        includeInBundle = false
    }

only this was changed , hopefully it maybe fixed now , otherwise only 1 more thing left to experiment with

**Comment by IzzySoft** [comment]
That has no effect on this diff (or on anything in the unsigned APK), only on the signing block. If you turned that on again (especially the former), your update will throw an error in the repo again, so please leave at least `includeInApk = false` (bundles do only affect PlayStore so far).

I see both are there and on `false` now. I've just run the test, v1.0.4 is not RB – and it's the same `.so` file again:

```diff
  -rw-r--r--  0.0 unx  1594196 b-   677509 defN 1981-01-01 01:01:02 8f55468e classes.dex
- -rw-r--r--  0.0 unx  5505952 b-  2177567 defN 1981-01-01 01:01:02 4bcfa472 lib/arm64-v8a/libapp.so
+ -rw-r--r--  0.0 unx  5505952 b-  2177567 defN 1981-01-01 01:01:02 8dabe765 lib/arm64-v8a/libapp.so
  -rw-r--r--  0.0 unx 10714752 b-  4991251 defN 1981-01-01 01:01:02 4a340eb2 lib/arm64-v8a/libflutter.so
```

Looking at the diffoscope results, exactly the same build-id stuff. If your build indeed uses `cmake: 3.31.1`, this could be the cause. And it's unfortunately nothing our builders can solve (we simply do not have that version). Currently, the only solution to that would be you switching to a supported cmake version, like 3.25.1 (which we'd have at Debian bookworm) or 3.28.3 (which we could use with Ubuntu noble). But everything newer than that would not be available to us (even the latest, non-LTS Ubuntu only has 3.30.3, and we don't use non-LTS here; at Debian only sid (unstable) would have it and we don't use unstable either).

So, not sure what updated your Linux toolchain to that (the "latest and greatest" from AUR maybe?) – could you "go back" with that to something "stable"?

**Comment by sidhant947** [comment]
Sorry for late response , due to some personal family issues i was busy, 

https://github.com/flutter/flutter/issues/139427

i saw a similiar issue with AGP so i opted to let flutter decide the ndk version instead of declaring it.

**Comment by IzzySoft** [comment]
Good news: v1.0.5 is RB again, congrats! And thanks to your including the Flutter version with `pubspec.yaml`, the recipe now can find the matching Flutter version itself, which then can find the matching NDK itself, so "for a while" we (hopefully) do not need to update the recipe here:

```yaml
build:
  - sed -r '/signingConfigs.release/d' -i android/app/build.gradle
  - FLUTTER_VERSION="$(sed -rn 's/\s*flutter:\s*\^([0-9.]*)/\1/p' pubspec.yaml)"
  - git clone -b $FLUTTER_VERSION https://github.com/flutter/flutter
  - export PUB_CACHE=$(pwd)/.pub-cache NO_SIGNING=true
  - flutter/bin/flutter config --no-analytics
  - flutter/bin/flutter pub get
  - flutter/bin/flutter build apk --release --split-per-abi --target-platform="android-arm64"
  - mv build/app/outputs/apk/release/app-arm64-v8a-release-unsigned.apk /outputs/unsigned.apk
```

Thanks again for your help and patience with this, much appreciated!

**Comment by IzzySoft** [comment]
Bad news: V1.0.9 failed again.

```diff
  -rw-r--r--  0.0 unx  5571488 b-  2248389 defN 1981-01-01 01:01:02 ac990891 lib/arm64-v8a/libapp.so
- -rw-r--r--  0.0 unx 10814192 b-  5027953 defN 1981-01-01 01:01:02 5e57d632 lib/arm64-v8a/libflutter.so
+ -rw-r--r--  0.0 unx 10833168 b-  5036083 defN 1981-01-01 01:01:02 938bf06e lib/arm64-v8a/libflutter.so
  -rw-r--r--  0.0 unx      770 b-      228 defN 1981-01-01 01:01:02 57efbdca assets/flutter_assets/AssetManifest.bin
```

Diffoscope: [diffoscope.zip](https://github.com/user-attachments/files/18323510/diffoscope.zip) (look at the section starting with "strings --all --bytes=8 {}" – to me the only one that gives some clue). Did you change something in your framework? I cannot look into your build setup, as you don't use Github actions for it. Maybe a different JDK? Shouldn't affect an `*.so`, but maybe indirectly, idk. NDK is more likely, but we didn't specify any before. You've used Flutter 3.27.0, as specified in your `pubspec.yaml`? Or did you maybe switch to 3.27.1 for your build?

**Comment by sidhant947** [comment]
Manjaro Auto Updated Flutter in Background , Fixed it now , Now it should be reproducible.

**Comment by IzzySoft** [comment]
Ouch. You *replaced* the APKs? Please don't do that again; they were already distributed via the repo – so while the RB might succeed now, the shield will not come up for this release as it's different APKs now.

Thanks for the fix, though: next release will then have the shield up again, hopefully :crossed_fingers:
